File: apps/android/VLinkerOBD/app/src/main/java/com/jslee1972/vlinkerobd/gps/GpsSpeedSource.kt
package com.jslee1972.vlinkerobd.gps

import kotlinx.coroutines.flow.StateFlow

/**
 * Phone-GPS-derived speed, kept separate from the vehicle's own OBD-reported speed so the two can
 * be shown side by side for comparison (a real-world sanity check on the vehicle's speed sensor).
 * Abstracted behind an interface — like [com.jslee1972.vlinkerobd.ble.BleObdClient] — so
 * [com.jslee1972.vlinkerobd.ui.DashboardViewModel] never touches Android's location APIs directly
 * and stays unit-testable with a fake.
 */
interface GpsSpeedSource {
    /** Current GPS-derived speed in km/h, or null before a first fix / while stopped listening. */
    val speedKph: StateFlow<Float?>

    /** Starts listening for location updates. A no-op if already started or permission is missing. */
    fun start()

    /** Stops listening; [speedKph] resets to null. */
    fun stop()
}

File: apps/ios/VLinkerOBD/GpsSpeedSource.swift
import CoreLocation
import Foundation

/// Phone GPS-derived speed, kept deliberately separate from the OBD-reported speed so the two can
/// be shown side-by-side as a sanity check on the vehicle's own speed sensor. Mirrors Android's
/// `GpsSpeedSource`/`AndroidGpsSpeedSource`.
///
/// Platform gap vs. Android: `LocationManager.GPS_PROVIDER` lets Android force GPS-chip-only
/// positioning with no WiFi/cell fusion; CoreLocation has no equivalent switch — `start()` here
/// requests the best available accuracy and lets iOS fuse whatever positioning sources it judges
/// best. This is a known, unavoidable platform difference (see shared/protocol-docs), not a bug.
protocol GpsSpeedSource: AnyObject {
    var speedKph: Float? { get }
    var onSpeedChange: ((Float?) -> Void)? { get set }
    func start()
    func stop()
}

final class CoreLocationGpsSpeedSource: NSObject, GpsSpeedSource, CLLocationManagerDelegate {
    private let manager = CLLocationManager()
    private(set) var speedKph: Float? {
        didSet { onSpeedChange?(speedKph) }
    }
    var onSpeedChange: ((Float?) -> Void)?

    override init() {
        super.init()
        manager.desiredAccuracy = kCLLocationAccuracyBest
        manager.distanceFilter = kCLDistanceFilterNone
        manager.delegate = self
    }

    func start() {
        let status = manager.authorizationStatus
        guard status == .authorizedWhenInUse || status == .authorizedAlways else { return }
        manager.startUpdatingLocation()
    }

    func stop() {
        manager.stopUpdatingLocation()
        speedKph = nil
    }

    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last else { return }
        // CLLocation.speed is -1 when the value is invalid (e.g. the first fix after cold start),
        // matching Android's `!hasSpeed()` guard against overwriting a good value with garbage.
        guard location.speed >= 0 else { return }
        speedKph = Float(location.speed) * 3.6
    }

    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        // Best-effort telemetry only — a failed GPS fix should not disrupt OBD polling.
    }
}
